File: contest-management-cs/contest-management-gui-cs/repository/database/RegistrationDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using contest_management_gui_cs.model;
using contest_management_gui_cs.utils;
using log4net;

namespace contest_management_gui_cs.repository
{
    public class RegistrationDbRepository : IRegistrationRepository
    {
        private readonly DbUtils _dbUtils;
        
        private static readonly ILog Log = LogManager.GetLogger("mylogger");

        public RegistrationDbRepository(string connectionName)
        {
            _dbUtils = new DbUtils(connectionName);
        }
        
        public IEnumerable<Registration> FindAll()
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select * from Registrations R
                inner join Participants P on P.id = R.participant_id
                inner join Competitions C on C.id = R.competition_id", conn);
            try
            {
                using var reader = cmd.ExecuteReader();
                var registrations = new List<Registration>();
                while (reader.Read())
                {
                    registrations.Add(new Registration
                    {
                        RegisteredParticipant = new Participant
                        {
                            Id = reader.GetInt64(2),
                            FirstName = reader.GetString(3),
                            LastName = reader.GetString(4),
                            Age = reader.GetInt32(5)
                        },
                        WhichCompetition = new Competition
                        {
                            Id = reader.GetInt64(6),
                            CompetitionType = reader.GetString(7),
                            AgeCategory = reader.GetString(8)
                        }
                    });
                }

                return registrations;
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return new List<Registration>();
        }

        public Registration Find(Tuple<long, long> id)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select * from Registrations R
                inner join Participants P on P.id = R.participant_id
                inner join Competitions C on C.id = R.competition_id
                where P.id = @participant_id and C.id = @competition_id",
                conn);
            cmd.Parameters.AddWithValue("@participant_id", id.Item1);
            cmd.Parameters.AddWithValue("@competition_id", id.Item2);
            try
            {
                using var reader = cmd.ExecuteReader();
                if (reader.Read())
                    return new Registration
                    {
                        RegisteredParticipant = new Participant
                        {
                            Id = reader.GetInt64(2),
                            FirstName = reader.GetString(3),
                            LastName = reader.GetString(4),
                            Age = reader.GetInt32(5)
                        },
                        WhichCompetition = new Competition
                        {
                            Id = reader.GetInt64(6),
                            CompetitionType = reader.GetString(7),
                            AgeCategory = reader.GetString(8)
                        }
                    };
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return default;
        }

        public void Add(Registration entity)
        {
            Log.TraceEntry(entity.ToString());

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"insert into Registrations (participant_id, competition_id) VALUES
                (@participant_id, @competition_id)",
                conn);
            cmd.Parameters.AddWithValue("@participant_id", entity.RegisteredParticipant.Id);
            cmd.Parameters.AddWithValue("@competition_id", entity.WhichCompetition.Id);
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB " + e);
            }
            finally
            {
                Log.TraceExit();
            }
        }

        public void Remove(Tuple<long, long> id)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"delete from Registrations
                where participant_id = @participant_id and competition_id = @competition_id", conn);
            cmd.Parameters.AddWithValue("@participant_id", id.Item1);
            cmd.Parameters.AddWithValue("@competition_id", id.Item2);
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB " + e);
            }
            finally
            {
                Log.TraceExit();
            }
        }

        public void Update(Registration entity)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Registration> FindByParticipant(long participantId)
        {
            Log.TraceEntry();

            var conn = _dbUtils.GetConnection();

            using var cmd = new SQLiteCommand(
                @"select * from Registrations R
                inner join Participants P on P.id = R.participant_id
                inner join Competitions C on C.id = R.competition_id
                where P.id = @participant_id",
                conn);
            cmd.Parameters.AddWithValue("@participant_id", participantId);
            try
            {
                using var reader = cmd.ExecuteReader();
                var registrations = new List<Registration>();
                while (reader.Read())
                    registrations.Add(new Registration
                    {
                        RegisteredParticipant = new Participant
                        {
                            Id = reader.GetInt64(2),
                            FirstName = reader.GetString(3),
                            LastName = reader.GetString(4),
                            Age = reader.GetInt32(5)
                        },
                        WhichCompetition = new Competition
                        {
                            Id = reader.GetInt64(6),
                            CompetitionType = reader.GetString(7),
                            AgeCategory = reader.GetString(8)
                        }
                    });
                
                return registrations;
            }
            catch (SQLiteException e)
            {
                Log.Error(e);
                Console.WriteLine("Error DB" + e);
            }
            finally
            {
                Log.TraceExit();
            }

            return default;
        }
    }
}

File: contest-management-client-server-java/ContestMgmtPersistence/src/main/java/contestmgmt/persistence/repository/jdbc/RegistrationDbRepository.java
package contestmgmt.persistence.repository.jdbc;

import contestmgmt.model.Competition;
import contestmgmt.model.Participant;
import contestmgmt.model.Registration;
import contestmgmt.model.Tuple;
import contestmgmt.persistence.repository.IRegistrationRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class RegistrationDbRepository implements IRegistrationRepository {
    private JdbcUtils dbUtils;

    public RegistrationDbRepository(Properties props) {
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Iterable<Registration> findAll() {
        Connection conn = dbUtils.getConnection();
        List<Registration> registrations = new ArrayList<>();

        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select * from Registrations as R " +
                        "inner join Participants as P on R.participant_id = P.id " +
                        "inner join Competitions as C on R.participant_id = C.id")) {
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    long participantId = result.getLong("participant_id");
                    String firstName = result.getString("first_name");
                    String lastName = result.getString("last_name");
                    int age = result.getInt("age");
                    long competitionId = result.getLong("competition_id");
                    String competitionType = result.getString("competition_type");
                    String ageCategory = result.getString("age_category");
                    Participant p = new Participant(firstName, lastName, age);
                    p.setId(participantId);
                    Competition c = new Competition(competitionType, ageCategory);
                    c.setId(competitionId);
                    Registration r = new Registration(p, c);
                    registrations.add(r);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return registrations;
    }

    @Override
    public Registration find(Tuple<Long, Long> longLongTuple) {
        Connection conn = dbUtils.getConnection();
        Registration r = null;
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select * from Registrations R " +
                        "inner join Participants P on R.participant_id = P.id " +
                        "inner join Competitions C on C.id = R.competition_id " +
                        "where participant_id = ? and competition_id = ?")) {
            prepStmt.setLong(1, longLongTuple.getLeft());
            prepStmt.setLong(2, longLongTuple.getRight());
            try (ResultSet result = prepStmt.executeQuery()) {
                result.next();
                long participantId = result.getLong("participant_id");
                String firstName = result.getString("first_name");
                String lastName = result.getString("last_name");
                int age = result.getInt("age");
                var p = new Participant(firstName, lastName, age);
                p.setId(participantId);

                long competitionId = result.getLong("competition_id");
                String competitionType = result.getString("competition_type");
                String ageCategory = result.getString("age_category");
                var c = new Competition(competitionType, ageCategory);
                c.setId(competitionId);

                r = new Registration(p, c);
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return r;
    }

    @Override
    public Tuple<Long, Long> add(Registration entity) {
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn.prepareStatement(
                        "insert into Registrations (participant_id, competition_id) " +
                                "values (?, ?)")) {
            prepStmt.setLong(1, entity.getParticipant().getId());
            prepStmt.setLong(2, entity.getCompetition().getId());
            int result = prepStmt.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB " + e);
        }
        return null;
    }

    @Override
    public void remove(Tuple<Long, Long> longLongTuple) {
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "delete from Registrations " +
                        "where participant_id = ? and competition_id = ?")) {
            prepStmt.setLong(1, longLongTuple.getLeft());
            prepStmt.setLong(2, longLongTuple.getRight());
            int result = prepStmt.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB " + e);
        }
    }

    @Override
    public void update(Registration entity) {

    }

    @Override
    public Iterable<Registration> findByParticipant(long participantId) {
        Connection conn = dbUtils.getConnection();
        List<Registration> registrations = new ArrayList<>();
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select * from Registrations R " +
                        "inner join Participants P on R.participant_id = P.id " +
                        "inner join Competitions C on C.id = R.competition_id " +
                        "where participant_id = ?")) {
            prepStmt.setLong(1, participantId);
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    long partId = result.getLong("participant_id");
                    String firstName = result.getString("first_name");
                    String lastName = result.getString("last_name");
                    int age = result.getInt("age");
                    var p = new Participant(firstName, lastName, age);
                    p.setId(participantId);

                    long competitionId = result.getLong("competition_id");
                    String competitionType = result.getString("competition_type");
                    String ageCategory = result.getString("age_category");
                    var c = new Competition(competitionType, ageCategory);
                    c.setId(competitionId);

                    Registration r = new Registration(p, c);
                    registrations.add(r);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return registrations;
    }
}
